### PR TITLE
store/tikv: abort an unsafe transaction subject to gc command

### PR DESCRIPTION
**Pull request by atmzhou**
Whenever a transaction fetches some data, we check whether the cached safepoint is fresh. If it is out-of-date, we abort the transaction. If the start timestamp of the transaction falls behind the cached safepoint, we also abort the transaction.

### CODE REVIEW

File: store/tikv/kv.go [hunk 1]
```diff
@@ -109,6 +112,93 @@ type tikvStore struct {
...
+		}
+		// Disable privilege check for gc worker session.
+		privilege.BindPrivilegeManager(s.spSession, nil)
+		log.Error("Exit createSPSession!\n")
+		return
+	}
+}
+
+func (s *tikvStore) saveUint64(key string, t uint64) error {
+	str := fmt.Sprintf("%v", t)
```
Comment (disksing) [inline store/tikv/kv.go]: use `strconv.FormatUint` instead.

File: store/tikv/gc_worker.go [hunk 2]
```diff
@@ -286,6 +289,7 @@ func RunGCJob(ctx goctx.Context, store kv.Storage, safePoint uint64, identifier
 }
 
 func (w *GCWorker) runGCJob(ctx goctx.Context, safePoint uint64) {
+	log.Error("[gc worker] enter runGCJob")
```
Comment (disksing) [inline store/tikv/gc_worker.go]: Remove useless debug logs, or change log level to Info/Debug.

File: store/tikv/gc_worker.go [hunk 3]
```diff
@@ -465,6 +469,10 @@ func resolveLocks(ctx goctx.Context, store *tikvStore, safePoint uint64, identif
 func doGC(ctx goctx.Context, store *tikvStore, safePoint uint64, identifier string) error {
 	gcWorkerCounter.WithLabelValues("do_gc").Inc()
 
+	store.gcWorker.saveUint64(gcSavedSafePoint, safePoint)
```
Comment (disksing) [inline store/tikv/gc_worker.go]: need check error.

File: store/tikv/gc_worker.go [hunk 4]
```diff
@@ -465,6 +469,10 @@ func resolveLocks(ctx goctx.Context, store *tikvStore, safePoint uint64, identif
...
+	log.Error("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
+	time.Sleep(100 * time.Second)
```
Comment (disksing) [inline store/tikv/gc_worker.go]: define a const variable for it.

File: store/tikv/gc_worker.go [hunk 5]
```diff
@@ -465,6 +469,10 @@ func resolveLocks(ctx goctx.Context, store *tikvStore, safePoint uint64, identif
 func doGC(ctx goctx.Context, store *tikvStore, safePoint uint64, identifier string) error {
 	gcWorkerCounter.WithLabelValues("do_gc").Inc()
 
+	store.gcWorker.saveUint64(gcSavedSafePoint, safePoint)
+	log.Error("[gc worker] %s start gc, safePoint: %v.", identifier, safePoint)
```
Comment (disksing) [inline store/tikv/gc_worker.go]: Do we need to move the log after sleep?

File: store/tikv/gc_worker.go [hunk 6]
```diff
@@ -587,6 +595,27 @@ func (w *GCWorker) saveTime(key string, t time.Time) error {
 	return errors.Trace(err)
 }
 
+func (w *GCWorker) saveUint64(key string, t uint64) error {
+	s := fmt.Sprintf("%v", t)
```
Comment (disksing) [inline store/tikv/gc_worker.go]: use `strconv.FormatUint` instead.

File: store/tikv/kv.go [hunk 7]
```diff
@@ -109,6 +112,93 @@ type tikvStore struct {
 	etcdAddrs    []string
 	mock         bool
 	enableGC     bool
+
+	safePoint uint64
+	spTime    time.Time
+	spSession tidb.Session // this is used to obtain safePoint from remote
+	spMutex   sync.Mutex   // this is used to update safePoint and spTime
```
Comment (disksing) [inline store/tikv/kv.go]: I think you want to use RWLock instead.

File: store/tikv/kv.go [hunk 8]
```diff
@@ -109,6 +112,93 @@ type tikvStore struct {
...
+	spMsg     chan string  // this is used to nofity when the store is closed
```
Comment (disksing) [inline store/tikv/kv.go]: use `chan struct{}` instead.

File: store/tikv/kv.go [hunk 9]
```diff
@@ -137,10 +232,47 @@ func (s *tikvStore) EtcdAddrs() []string {
...
+					newSafePoint, err := s.loadUint64(gcSavedSafePoint)
+					if err == nil {
+						s.spMutex.Lock()
+						s.safePoint = newSafePoint
+						s.spTime = time.Now()
+						s.spMutex.Unlock()
+						log.Error("[safepoint load OK]\n")
+					} else {
+						s.spMutex.Lock()
+						s.spTime = time.Now()
```
Comment (disksing) [inline store/tikv/kv.go]: I don't think we need to update `spTime` when request failed.

File: store/tikv/kv.go [hunk 10]
```diff
@@ -137,10 +232,47 @@ func (s *tikvStore) EtcdAddrs() []string {
...
+						s.spMutex.Unlock()
+						log.Error("[safepoint load error]\n")
+						repeatbreak = true
+					}
+					time.Sleep(5 * time.Second) // this is configurable
```
Comment (disksing) [inline store/tikv/kv.go]: define a const for it.

File: vendor [hunk 11]
```diff
@@ -0,0 +1 @@
+_vendor/src/
```
Comment (disksing) [inline vendor]: remove this unnecessary file.

File: store/tikv/safepoint_test.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (disksing) [inline store/tikv/safepoint_test.go]: 2016 -> 2017

File: store/tikv/gc_worker.go [hunk 13]
```diff
@@ -36,17 +36,48 @@ import (
...
+			cachedSafePoint, cacheErr := w.loadSafePoint(gcSavedSafePoint)
+			if cacheErr == nil {
+				w.store.spMutex.Lock()
+				w.store.safePoint = cachedSafePoint
+				w.store.spTime = spCachedTime
+				w.store.spMutex.Unlock()
+			} else {
+				log.Error("[safepoint] read error:", cacheErr)
+			}
+			time.Sleep(5 * time.Second) // this is configurable
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply store/tikv/gc_worker.go]: You should write code like this:
```
ticker := time.NewTicker()
select {
    case <-w.store.spMsg:
    case <-ticker.C:
        ...
}
```
Because when you `Sleep()`, w.store.spMsg signal will be miss.

File: store/tikv/gc_worker.go [hunk 14]
```diff
@@ -36,17 +36,48 @@ import (
...
+	spSession tidb.Session
+
+	gcIsRunning bool
 	lastFinish  time.Time
 	cancel      goctx.CancelFunc
 	done        chan error
 }
 
+// StartSafePointChecker triggers SafePoint Checker on each tidb
+func (w *GCWorker) StartSafePointChecker() {
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: I doubt operate safe point in two different worker can't guarantee data consistency.

Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: I doubt that operating safe point in two different workers can't guarantee data consistency.

File: store/tikv/gc_worker.go [hunk 15]
```diff
@@ -178,7 +227,8 @@ func (w *GCWorker) storeIsBootstrapped() bool {
 
 // Leader of GC worker checks if it should start a GC job every tick.
 func (w *GCWorker) leaderTick(ctx goctx.Context) error {
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Please rebase master to eliminate unrelated changes.

File: store/tikv/gc_worker.go [hunk 16]
```diff
@@ -465,6 +512,9 @@ func resolveLocks(ctx goctx.Context, store *tikvStore, safePoint uint64, identif
 func doGC(ctx goctx.Context, store *tikvStore, safePoint uint64, identifier string) error {
 	gcWorkerCounter.WithLabelValues("do_gc").Inc()
 
+	store.gcWorker.saveSafePoint(gcSavedSafePoint, safePoint)
+	time.Sleep(100 * time.Second)
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: why sleep here?

File: store/tikv/gc_worker.go [hunk 17]
```diff
@@ -691,10 +754,17 @@ func NewMockGCWorker(store kv.Storage) (*MockGCWorker, error) {
 		uuid:        strconv.FormatUint(ver.Ver, 16),
 		desc:        fmt.Sprintf("host:%s, pid:%d, start at %s", hostName, os.Getpid(), time.Now()),
 		store:       store.(*tikvStore),
-		gcIsRunning: 0,
+		gcIsRunning: false,
 		lastFinish:  time.Now(),
 		done:        make(chan error),
 	}
+	worker.session, err = tidb.CreateSession(worker.store)
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Several goroutine use the same session at the same time may data race.

File: store/tikv/kv.go [hunk 18]
```diff
@@ -109,6 +110,26 @@ type tikvStore struct {
...
 	enableGC     bool
+
+	safePoint uint64
+	spTime    time.Time
+	spSession tidb.Session  // this is used to obtain safePoint from remote
+	spMutex   sync.RWMutex  // this is used to update safePoint and spTime
+	spMsg     chan struct{} // this is used to nofity when the store is closed
+}
+
+func (s *tikvStore) CheckVisibility() (uint64, error) {
```
Comment (tiancaiamao) [inline store/tikv/kv.go]: This function name is not consistent with its return values.
`CheckXXX` should return bool.

File: store/tikv/safepoint_test.go [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+	return keys
+}
+
+func (s *testSafePointSuite) TestSafePoint(c *C) {
+	txn := s.beginTxn(c)
+	for i := 0; i < 10; i++ {
+		seterr := txn.Set(encodeKey(s.prefix, s08d("key", i)), valueBytes(i))
+		c.Assert(seterr, IsNil)
+	}
+	commiterr := txn.Commit()
```
Comment (tiancaiamao) [inline store/tikv/safepoint_test.go]: s/commiterr/commitErr

File: store/tikv/safepoint_test.go [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+	c.Assert(commiterr, IsNil)
+
+	// for txn get
+	txn2 := s.beginTxn(c)
+	_, geterr := txn2.Get(encodeKey(s.prefix, s08d("key", 0)))
+	c.Assert(geterr, IsNil)
+
+	for {
+		s.gcWorker.saveSafePoint(gcSavedSafePoint, txn2.startTS+10)
+		newSafePoint, loaderr := s.gcWorker.loadSafePoint(gcSavedSafePoint)
```
Comment (tiancaiamao) [inline store/tikv/safepoint_test.go]: s/loaderr/loadErr, or just err,  for short.

File: store/tikv/safepoint_test.go [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+		if loaderr == nil {
+			s.store.spMutex.Lock()
+			s.store.safePoint = newSafePoint
+			s.store.spTime = time.Now()
+			s.store.spMutex.Unlock()
+			break
+		} else {
+			time.Sleep(5 * time.Second)
```
Comment (tiancaiamao) [inline store/tikv/safepoint_test.go]: Sleep so long make the CI too slow...

File: store/tikv/safepoint_test.go [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+		}
+	}
+
+	_, geterr2 := txn2.Get(encodeKey(s.prefix, s08d("key", 0)))
+	c.Assert(geterr2, NotNil)
+
+	// for txn seek
+	txn3 := s.beginTxn(c)
+	for {
```
Comment (tiancaiamao) [inline store/tikv/safepoint_test.go]: Don't repeat yourself when you find some code looks familiar.

File: store/tikv/safepoint_test.go [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+			s.store.spTime = time.Now()
+			s.store.spMutex.Unlock()
+			break
+		} else {
+			time.Sleep(5 * time.Second)
+		}
+	}
+
+	_, seekerr := txn3.Seek(encodeKey(s.prefix, ""))
+	c.Assert(seekerr, NotNil)
```
Comment (tiancaiamao) [inline store/tikv/safepoint_test.go]: Check error.

File: store/tikv/scan.go [hunk 24]
```diff
@@ -168,6 +168,17 @@ func (s *Scanner) getData(bo *Backoffer) error {
 			return errors.Trace(errBodyMissing)
 		}
 
+		var safePoint uint64
```
Comment (tiancaiamao) [inline store/tikv/scan.go]: Could this check logic move to just one place?

File: store/tikv/gc_worker.go [hunk 25]
```diff
@@ -43,10 +44,40 @@ type GCWorker struct {
...
+	go func() {
+		for {
+			select {
+			case <-w.store.spMsg:
+				return
+			default:
+				spCachedTime := time.Now()
+				cachedSafePoint, err := w.loadSafePoint(gcSavedSafePoint)
+				if err == nil {
+					w.store.spMutex.Lock()
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: I prefer to create an independent function in `store` to set `safePoint` and `spTime`

File: store/tikv/gc_worker.go [hunk 26]
```diff
@@ -101,11 +145,15 @@ var gcVariableComments = map[string]string{
...
 	gcLifeTimeKey:    "All versions within life time will not be collected by GC, at least 10m, in Go format.",
 	gcSafePointKey:   "All versions after safe point can be accessed. (DO NOT EDIT)",
+	gcSavedSafePoint: "Saved SafePoint",
 }
 
-func (w *GCWorker) start(ctx goctx.Context) {
+func (w *GCWorker) start(ctx goctx.Context, wg *sync.WaitGroup) {
 	log.Infof("[gc worker] %s start.", w.uuid)
+
+	w.session = createSession(w.store)
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: Why we need to `createSession` again after L101?

File: store/tikv/gc_worker.go [hunk 27]
```diff
@@ -129,15 +177,15 @@ func (w *GCWorker) start(ctx goctx.Context) {
 
 func createSession(store kv.Storage) tidb.Session {
 	for {
-		session, err := tidb.CreateSession(store)
-		if err != nil {
-			log.Warnf("[gc worker] create session err: %v", err)
+		retSession, sessionErr := tidb.CreateSession(store)
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: Why rename them?

File: store/tikv/gc_worker.go [hunk 28]
```diff
@@ -43,10 +44,40 @@ type GCWorker struct {
...
+			case <-w.store.spMsg:
+				return
+			default:
+				spCachedTime := time.Now()
+				cachedSafePoint, err := w.loadSafePoint(gcSavedSafePoint)
+				if err == nil {
+					w.store.spMutex.Lock()
+					w.store.safePoint = cachedSafePoint
+					w.store.spTime = spCachedTime
+					w.store.spMutex.Unlock()
```
Comment (disksing) [inline store/tikv/gc_worker.go]: why not move `spMutex` `safePoint` `spTime` `spMsg` to GCWorker too?

File: store/tikv/kv.go [hunk 29]
```diff
@@ -109,6 +110,31 @@ type tikvStore struct {
...
+
+func (s *tikvStore) CheckVisibility(startTime uint64) error {
+	s.spMutex.RLock()
+	cachedSafePoint := s.safePoint
+	cachedTime := s.spTime
+	s.spMutex.RUnlock()
+	diff := time.Since(cachedTime)
+
+	// the magic number "5" is used to compensate the inaccuracy of CPU clock
+	if diff > (gcSafePointCacheInterval - 5) {
```
Comment (disksing) [inline store/tikv/kv.go]: define a const for the magic number.

Comment (disksing) [reply store/tikv/kv.go]: 5 seems too small, how about `time.Second`?

**Comment by disksing** [comment]
LGTM.

File: store/tikv/gc_worker.go [hunk 30]
```diff
@@ -43,10 +44,37 @@ type GCWorker struct {
...
+	sessionLock sync.Mutex
+}
+
+// StartSafePointChecker triggers SafePoint Checker on each tidb
+func (w *GCWorker) StartSafePointChecker() {
+	w.spsession = createSession(w.store)
+	go func() {
+		for {
+			select {
+			case <-w.store.spMsg:
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Don't use sleep, use 
```
ticker := time.NewTicker()
select {
    case <-w.store.spMsg:
    case <-ticker.C:
}
```
When you sleep, you'll miss signal from `w.store.spMsg`

File: store/tikv/gc_worker.go [hunk 31]
```diff
@@ -43,10 +44,37 @@ type GCWorker struct {
...
+				return
+			default:
+				spCachedTime := time.Now()
+				cachedSafePoint, err := w.loadSafePoint(gcSavedSafePoint)
+				if err == nil {
+					w.store.UpdateSPCache(cachedSafePoint, spCachedTime)
+				} else {
+					log.Warnf("There is no safepoint or the read fails")
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Those two cases is different, I think just read fails  should return error,
no safepoint equals that safepoint = 0
You should distinguish them and print error information.

File: store/tikv/gc_worker.go [hunk 32]
```diff
@@ -101,11 +143,14 @@ var gcVariableComments = map[string]string{
 	gcRunIntervalKey: "GC run interval, at least 10m, in Go format.",
 	gcLifeTimeKey:    "All versions within life time will not be collected by GC, at least 10m, in Go format.",
 	gcSafePointKey:   "All versions after safe point can be accessed. (DO NOT EDIT)",
+	gcSavedSafePoint: "Saved SafePoint",
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Saved safe point.

File: store/tikv/gc_worker.go [hunk 33]
```diff
@@ -465,6 +510,9 @@ func resolveLocks(ctx goctx.Context, store *tikvStore, safePoint uint64, identif
 func doGC(ctx goctx.Context, store *tikvStore, safePoint uint64, identifier string) error {
 	gcWorkerCounter.WithLabelValues("do_gc").Inc()
 
+	store.gcWorker.saveSafePoint(gcSavedSafePoint, safePoint)
+	time.Sleep(gcSafePointCacheInterval)
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Add comment herer.
I think it's better to getTS() again and ensure there are at least `gcSafePointCacheInterval` escaped, depending on local clock is unreliable.

File: store/tikv/gc_worker.go [hunk 34]
```diff
@@ -43,10 +44,37 @@ type GCWorker struct {
 	lastFinish  time.Time
 	cancel      goctx.CancelFunc
 	done        chan error
+
+	session tidb.Session
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Data race may exists, I see both `doGC` and `StartSafePointChecker` goroutine call `loadSafePoint` related function, and they both use this session?

File: store/tikv/scan.go [hunk 35]
```diff
@@ -168,6 +168,12 @@ func (s *Scanner) getData(bo *Backoffer) error {
 			return errors.Trace(errBodyMissing)
 		}
 
+		checkerr := s.snapshot.store.CheckVisibility(s.startTS())
+
```
Comment (tiancaiamao) [inline store/tikv/scan.go]: useless empty line

File: store/tikv/snapshot.go [hunk 36]
```diff
@@ -78,6 +78,13 @@ func (s *tikvSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+
+	checkerr := s.store.CheckVisibility(s.version.Ver)
+
```
Comment (tiancaiamao) [inline store/tikv/snapshot.go]: ditto

File: store/tikv/txn.go [hunk 37]
```diff
@@ -75,6 +75,13 @@ func (txn *tikvTxn) Get(k kv.Key) ([]byte, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+
+	checkerr := txn.store.CheckVisibility(txn.startTS)
+
```
Comment (tiancaiamao) [inline store/tikv/txn.go]: ditto

File: store/tikv/kv.go [hunk 38]
```diff
@@ -109,6 +110,37 @@ type tikvStore struct {
...
+	s.spTime = cachedTime
+	s.spMutex.Unlock()
+}
+
+func (s *tikvStore) CheckVisibility(startTime uint64) error {
+	s.spMutex.RLock()
+	cachedSafePoint := s.safePoint
+	cachedTime := s.spTime
+	s.spMutex.RUnlock()
+	diff := time.Since(cachedTime)
```
Comment (tiancaiamao) [inline store/tikv/kv.go]: It's possible that `SafePointChecker` can't load data in 10s...
Then this check make read data from TiKV totally unavailable, what's worse, 
if CheckVisibility fail, we can't read and update safepoint, 
if we can't update safepoint, CheckVisibility fail again...

File: store/tikv/safepoint_test.go [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+
+	// for txn get
+	txn2 := s.beginTxn(c)
+	_, err = txn2.Get(encodeKey(s.prefix, s08d("key", 0)))
+	c.Assert(err, IsNil)
+
+	s.waitUntilErrorPlugIn(txn2.startTS)
+
+	_, geterr2 := txn2.Get(encodeKey(s.prefix, s08d("key", 0)))
+	c.Assert(geterr2, NotNil)
```
Comment (tiancaiamao) [inline store/tikv/safepoint_test.go]: Check error detail, not just error not nil.

File: store/tikv/gc_worker.go (hunk 30 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: OK, I have revised.

File: store/tikv/gc_worker.go (hunk 32 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: OK, this is no longer needed.

File: store/tikv/kv.go (hunk 38 above)
Comment (atmzhou) [reply store/tikv/kv.go]: Good Insight! I have resorted to ETCD for storing data.
Thanks!

File: store/tikv/scan.go (hunk 35 above)
Comment (atmzhou) [reply store/tikv/scan.go]: OK, I have delete the empty line. Thanks~

File: store/tikv/snapshot.go (hunk 36 above)
Comment (atmzhou) [reply store/tikv/snapshot.go]: Yes, I have revised it. Thanks~

File: store/tikv/txn.go (hunk 37 above)
Comment (atmzhou) [reply store/tikv/txn.go]: Revised also, Thanks~

File: store/tikv/safepoint_test.go (hunk 39 above)
Comment (atmzhou) [reply store/tikv/safepoint_test.go]: Thanks, I have revised.

File: store/tikv/gc_worker.go (hunk 33 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Good Insight! Thanks~

File: store/tikv/gc_worker.go (hunk 31 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Thanks, I have revised it.

File: store/tikv/gc_worker.go (hunk 34 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Thanks, it has been changed.

File: store/tikv/2pc.go [hunk 40]
```diff
@@ -430,7 +430,7 @@ func (c *twoPhaseCommitter) commitSingleBatch(bo *Backoffer, batch batchKeys) er
 	// Under this circumstance,  we can not declare the commit is complete (may lead to data lost), nor can we throw
 	// an error (may lead to the duplicated key error when upper level restarts the transaction). Currently the best
 	// solution is to populate this error and let upper layer drop the connection to the corresponding mysql client.
-	isPrimary := bytes.Equal(batch.keys[0], c.primary())
+	isPrimary := bytes.Compare(batch.keys[0], c.primary()) == 0
```
Comment (disksing) [inline store/tikv/2pc.go]: why change it?

File: store/tikv/2pc_test.go [hunk 41]
```diff
@@ -42,7 +42,10 @@ func (s *testCommitterSuite) SetUpTest(c *C) {
 	mvccStore := mocktikv.NewMvccStore()
 	client := mocktikv.NewRPCClient(s.cluster, mvccStore)
 	pdCli := &codecPDClient{mocktikv.NewPDClient(s.cluster)}
-	store, err := newTikvStore("mock-tikv-store", pdCli, client, false)
+	spkv := &MockSafePointKV{
+		store: make(map[string]string),
+	}
```
Comment (disksing) [inline store/tikv/2pc_test.go]: I think it would be better to add a `newMockSafepointKV` function.

File: store/tikv/gc_worker.go [hunk 42]
```diff
@@ -582,13 +707,48 @@ func (w *GCWorker) checkLeader() (bool, error) {
 	return false, nil
 }
 
-func (w *GCWorker) saveTime(key string, t time.Time) error {
-	err := w.saveValueToSysTable(key, t.Format(gcTimeFormat))
+func (w *GCWorker) saveSafePoint(key string, t uint64) error {
+	w.checkerLock.Lock()
+	defer w.checkerLock.Unlock()
```
Comment (disksing) [inline store/tikv/gc_worker.go]: etcd client supports accessing concurrently. So it's better to remove the mutex. btw, don't forget to make sure `MockSafePointKV` is thread safe too.

File: store/tikv/gc_worker.go [hunk 43]
```diff
@@ -43,10 +45,92 @@ type GCWorker struct {
...
+func (w *EtcdSafePointKV) Get(k string) (string, error) {
+	ctx, cancel := goctx.WithTimeout(goctx.Background(), time.Second*5)
+	resp, err := w.cli.Get(ctx, k)
+	cancel()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	for _, ev := range resp.Kvs {
+		return string(ev.Value), nil
+	}
```
Comment (disksing) [inline store/tikv/gc_worker.go]: How about
```
if len(resp.Kvs) > 0 {
  return string(resp.Kvs[0].Value), nil
}
```

File: store/tikv/2pc.go (hunk 40 above)
Comment (atmzhou) [reply store/tikv/2pc.go]: I do not make the change, I will rebase my code.

File: store/tikv/gc_worker.go (hunk 42 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Thanks, I have revised accordingly.

File: store/tikv/gc_worker.go (hunk 43 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Ok， I have revised.

File: store/tikv/2pc_test.go (hunk 41 above)
Comment (atmzhou) [reply store/tikv/2pc_test.go]: Ok, I have revised.

File: store/tikv/gc_worker.go (hunk 42 above)
Comment (tiancaiamao) [reply store/tikv/gc_worker.go]: B.T.W, the mock client is not thread safe for concurrently accessing...

File: store/tikv/gc_worker.go [hunk 44]
```diff
@@ -582,13 +707,48 @@ func (w *GCWorker) checkLeader() (bool, error) {
...
+func (w *GCWorker) saveSafePoint(key string, t uint64) error {
+	w.checkerLock.Lock()
+	defer w.checkerLock.Unlock()
+	s := strconv.FormatUint(t, 10)
+	err := w.store.kv.Put(gcSavedSafePoint, s)
+	if err != nil {
+		log.Error("save safepoint failed:", err)
+		return errors.Trace(err)
+	}
+	//log.Error("write safe point:", s)
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: Clean up

File: store/tikv/gc_worker.go [hunk 45]
```diff
@@ -582,13 +707,48 @@ func (w *GCWorker) checkLeader() (bool, error) {
...
+
+	if str == "" {
+		return 0, nil
+	}
+
+	t, err := strconv.ParseUint(str, 10, 64)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	//log.Error("read safe point:", t)
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: ditto

File: store/tikv/gc_worker.go [hunk 46]
```diff
@@ -43,10 +45,92 @@ type GCWorker struct {
...
+	Get(k string) (string, error)
+}
+
+// MockSafePointKV implements SafePointKV at mock test
+type MockSafePointKV struct {
+	store map[string]string
+}
+
+// Put implements the Put method for SafePointKV
+func (w *MockSafePointKV) Put(k string, v string) error {
```
[the remark this answers is not in the document]
Comment (atmzhou) [reply store/tikv/gc_worker.go]: But in the old version, it is protected by lock in its caller

[the remark this answers is not in the document]
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Now, it has been revised.

File: store/tikv/gc_worker.go [hunk 47]
```diff
@@ -43,10 +45,92 @@ type GCWorker struct {
...
+
+// StartSafePointChecker triggers SafePoint Checker on each tidb
+func (w *GCWorker) StartSafePointChecker() {
+	go func() {
+		ticker := time.NewTicker(gcSafePointUpdateInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				spCachedTime := time.Now()
```
[the remark this answers is not in the document]
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Thanks for this note!

File: store/tikv/gc_worker.go [hunk 48]
```diff
@@ -465,6 +571,20 @@ func resolveLocks(ctx goctx.Context, store *tikvStore, safePoint uint64, identif
...
+	// We want to make sure to delay enough time even if the Sleep is disrupted
+	{
+		startTime := time.Now()
+		leftTime := gcSafePointCacheInterval
+		for {
+			time.Sleep(leftTime)
+			if time.Since(startTime) > gcSafePointCacheInterval {
+				break
+			}
+			leftTime = leftTime - time.Since(startTime)
```
[the remark this answers is not in the document]
Comment (atmzhou) [reply store/tikv/gc_worker.go]: OK, I have used this code.
Compared to my code, the sleep time may double.

File: store/tikv/gc_worker.go (hunk 44 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: OK, Thanks~

File: store/tikv/gc_worker.go (hunk 45 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: OK, Thanks~

File: store/tikv/gc_worker.go [hunk 49]
```diff
@@ -43,10 +45,110 @@ type GCWorker struct {
...
+		ticker := time.NewTicker(gcSafePointUpdateInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case spCachedTime := <-ticker.C:
+				cachedSafePoint, err := w.loadSafePoint(gcSavedSafePoint)
+				if err == nil {
+					w.store.UpdateSPCache(cachedSafePoint, spCachedTime)
+				} else {
+					log.Error("The read fails:", errors.ErrorStack(err))
```
Comment (tiancaiamao) [inline store/tikv/gc_worker.go]: I think we should at least retry 3~5 times if error happens, otherwise `spCachedTime` won't be updated for at least 10s,
during which `CheckVisibility` would fail and service is unavailable.

If `CheckVisibility` fail, would us backoff that operation? @disksing 
And if `spCachedTime` is not updated, backoff won't work.

**Comment by tiancaiamao** [comment]
LGTM

File: store/tikv/gc_worker.go [hunk 50]
```diff
@@ -43,10 +45,111 @@ type GCWorker struct {
...
+		return string(resp.Kvs[0].Value), nil
+	}
+	return "", nil
+}
+
+// StartSafePointChecker triggers SafePoint Checker on each tidb
+func (w *GCWorker) StartSafePointChecker() {
+	go func() {
+		for {
+			d := gcSafePointUpdateInterval
```
Comment (disksing) [inline store/tikv/gc_worker.go]: this line should go before `for`.

**Comment by disksing** [comment]
LGTM. PTAL @AndreMouche

File: store/tikv/gc_worker.go [hunk 51]
```diff
@@ -43,10 +45,111 @@ type GCWorker struct {
...
+	w.store[k] = v
+	return nil
+}
+
+// Get implements the Get method for SafePointKV
+func (w *MockSafePointKV) Get(k string) (string, error) {
+	w.mockLock.RLock()
+	defer w.mockLock.RUnlock()
+	elem, ok := w.store[k]
+	if !ok {
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: I prefer to return` w.store[k], nil ` directly

File: store/tikv/gc_worker.go [hunk 52]
```diff
@@ -43,10 +45,111 @@ type GCWorker struct {
...
+// Put implements the Put method for SafePointKV
+func (w *MockSafePointKV) Put(k string, v string) error {
+	w.mockLock.Lock()
+	defer w.mockLock.Unlock()
+	w.store[k] = v
+	return nil
+}
+
+// Get implements the Get method for SafePointKV
+func (w *MockSafePointKV) Get(k string) (string, error) {
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: `error` seems redundant since it always return nil?

File: store/tikv/gc_worker.go [hunk 53]
```diff
@@ -43,10 +45,111 @@ type GCWorker struct {
...
+	}
+	return &EtcdSafePointKV{cli: etcdCli}, nil
+}
+
+// Put implements the Put method for SafePointKV
+func (w *EtcdSafePointKV) Put(k string, v string) error {
+	ctx, cancel := goctx.WithTimeout(goctx.Background(), time.Second*5)
+	_, err := w.cli.Put(ctx, k, v)
+	cancel()
+	if err != nil {
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: You can return `errors.Trace(err)` directly here

File: store/tikv/gc_worker.go [hunk 54]
```diff
@@ -103,9 +219,14 @@ var gcVariableComments = map[string]string{
 	gcSafePointKey:   "All versions after safe point can be accessed. (DO NOT EDIT)",
 }
 
-func (w *GCWorker) start(ctx goctx.Context) {
+func (w *GCWorker) start(ctx goctx.Context, wg *sync.WaitGroup) {
 	log.Infof("[gc worker] %s start.", w.uuid)
+
+	w.session = createSession(w.store)
+	log.Error("GC Worker creates session")
```
Comment (AndreMouche) [inline store/tikv/gc_worker.go]: Why log Error?

File: store/tikv/gc_worker.go (hunk 52 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: MockSafePointKV implements SafePointKV.
SafePointKV is also implemented by EtcdSafePointKV, which may return not nil.
So, the interface should be compatible with both implementations.

File: store/tikv/gc_worker.go (hunk 51 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: the checker may try to read when no gc worker saves safepoint. In this case, we should check the ok variable.

File: store/tikv/gc_worker.go (hunk 53 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Thanks~

File: store/tikv/gc_worker.go (hunk 54 above)
Comment (atmzhou) [reply store/tikv/gc_worker.go]: Thanks~I will remove this.
